runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp

```diff
@@ -39,3 +39,31 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
    TR_ASSERT_FATAL(false, "This helper implements platform specific code for Fieldwatch, which is currently not supported on ARM64 platforms.\n");
    }
 	
+TR::Register *J9::ARM64::TreeEvaluator::arraylengthEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(cg->fe());
+   // ldrw R, [B + contiguousSize]        ; Load contiguous array length
+   // cmp R, 0                            ; If 0, must be a discontiguous array
+   // ldrw R, [B + discontiguousSize]     ; Load discontiguous array length
+   //b done
+   TR::Register *objectReg = cg->evaluate(node->getFirstChild());
+   TR::Register *lengthReg = cg->allocateRegister();
+   TR::Register *condReg = cg->allocateRegister(TR_CCR);
+   TR::LabelSymbol *doneLabel = generateLabelSymbol(cg);
+
+   TR::MemoryReference *contiguousArraySizeMR = new (cg->trHeapMemory()) TR::MemoryReference(objectReg, fej9->getOffsetOfContiguousArraySizeField(), 4, cg);
+   TR::MemoryReference *discontiguousArraySizeMR = new (cg->trHeapMemory()) TR::MemoryReference(objectReg, fej9->getOffsetOfDiscontiguousArraySizeField(), 4, cg);
+
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, lengthReg, contiguousArraySizeMR);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::ccmpimmw, node, condReg, lengthReg, 0);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, discontiguousArrayLabel, condReg, TR::CC_EQ);
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, doneLabel);
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, lengthReg, discontiguousArraySizeMR);
+   generateLabelInstruction(cg, TR::InstOpCode::b, node, doneLabel);
```

Comment:
You have to be careful about introducing control flow in register sequences because of the interaction with the register assigner (which operates linearly).  You don't want spilling to occur on one path and not the other.  Generally, the way you work around that is to force spilling at a controlled point, namely the merge point label.  The way you do that is to add a `NoReg` register dependency to this label for the registers being referenced (in your case `lengthReg`).